### PR TITLE
some progress on batchnorms (draft)

### PR DESCRIPTION
**Pull request by marcelbischoff**

### CODE REVIEW

File: tinygrad/nn.py [hunk 1]
```diff
@@ -1,23 +1,37 @@
...
     self.running_var = Tensor.ones(sz, requires_grad=False)
     self.num_batches_tracked = Tensor.zeros(1, requires_grad=False)
 
   def __call__(self, x):
-    # TODO: use tinyops for this
-    # mean op needs to support the axis argument before we can do this
-    #self.running_mean.data = x.data.mean(axis=(0,2,3))
-    #self.running_var.data = ((x - self.running_mean.reshape(shape=[1, -1, 1, 1]))**self.two).data.mean(axis=(0,2,3))
+    [bs, sz], m = x.shape[:2], x.shape[2]*x.shape[3]
+    div =  Tensor([1/bs/m],  gpu=x.gpu, requires_grad=False)   #we are reimplementing means by contracting with 1's
```
Comment (geohot) [inline tinygrad/nn.py:18]: We need to just add axis support to sum (then mean is almost free)

File: tinygrad/nn.py [hunk 2]
```diff
@@ -1,23 +1,37 @@
...
+    ccol =  Tensor.ones(m,1, gpu=x.gpu, requires_grad=False)   #we are reimplementing means by contracting with 1's
+    batch_mean = crow.dot(x.reshape(shape=[bs,sz*m])).reshape(shape=[sz,m]).dot(ccol).reshape(shape=[sz]).mul(div)
+    y = pow((x - self.running_mean.reshape(shape=[1, -1, 1, 1])).reshape(shape=[-1]), self.two)
+    #y = (x - self.running_mean.reshape(shape=[1, -1, 1, 1])).mul(x - self.running_mean.reshape(shape=[1, -1, 1, 1]))#, self.two) 
+    batch_var = crow.dot(y.reshape(shape=[bs,sz*m])).reshape(shape=[sz,m]).dot(ccol).reshape(shape=[sz]).mul(div)
+    if self.track_running_stats: #needs momentum
+        self.running_mean = self.running_mean.mul(self.num_batches_tracked).add(batch_mean)
+        self.running_var = self.running_var.mul(self.num_batches_tracked).add(batch_var)
+        self.num_batches_tracked = self.num_batches_tracked.add(Tensor.ones(1, requires_grad=False))
+        self.running_mean = self.running_mean.mul(self.num_batches_tracked).div(self.num_batches_tracked)
```
Comment (geohot) [inline tinygrad/nn.py:29]: Unless I'm reading it wrong, I don't think you want a second mul here.

**Comment by marcelbischoff** [comment]
It is kinda learning  `TRANSFER=1 examples/train_efficientnet.py`. I think loading the weights on GPU is messed up.

**Comment by marcelbischoff** [comment]
close?
